sling/task/frames.cc

```diff
@@ -0,0 +1,173 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "sling/task/frames.h"
+
+#include "sling/base/logging.h"
+#include "sling/frame/encoder.h"
+#include "sling/frame/decoder.h"
+#include "sling/frame/object.h"
+#include "sling/frame/reader.h"
+#include "sling/frame/store.h"
+#include "sling/frame/wire.h"
+#include "sling/stream/file.h"
+#include "sling/stream/memory.h"
+#include "sling/task/task.h"
+
+namespace sling {
+namespace task {
+
+void FrameProcessor::Start(Task *task) {
+  // Create commons store.
+  commons_ = new Store();
+
+  // Load commons store from file.
+  for (Binding *binding : task->GetInputs("commons")) {
+    LoadStore(commons_, binding->resource());
+  }
+
+  // Get output channel (optional).
+  output_ = task->GetSink("output");
+
+  // Bind names.
+  CHECK(names_.Bind(commons_));
+  InitCommons(task);
+
+  // Start up processor.
+  Startup(task);
+
+  // Freeze commons store.
+  commons_->Freeze();
+
+  // Update statistics for common store.
+  MemoryUsage usage;
+  commons_->GetMemoryUsage(&usage, true);
+  task->GetCounter("commons_memory")->Increment(usage.memory_used());
+  task->GetCounter("commons_handles")->Increment(usage.used_handles());
+  task->GetCounter("commons_symbols")->Increment(usage.num_symbols());
+  task->GetCounter("commons_gcs")->Increment(usage.num_gcs);
+  task->GetCounter("commons_gctime")->Increment(usage.gc_time);
+
+  // Get counters for frame stores.
+  frame_memory_ = task->GetCounter("frame_memory");
+  frame_handles_ = task->GetCounter("frame_handles");
+  frame_symbols_ = task->GetCounter("frame_symbols");
+  frame_gcs_ = task->GetCounter("frame_gcs");
+  frame_gctime_ = task->GetCounter("frame_gctime");
+}
+
+void FrameProcessor::Receive(Channel *channel, Message *message) {
```

Comment:
Yes, the message content is always plain byte arrays. I did look into having "typed messages", but I decided against it because it makes lifetime management of messages much more complicated. We don't have a "Framer pipeline" yet, but when we add this, we can send the frames between the framers without having to encode and decode between the framers.